dv/uvm/core_ibex/collect_results.py

```diff
@@ -4,78 +4,217 @@
+
+def build_broken_test_run_result(err: str) -> TestRunResult:
+    return TestRunResult(
+            name = 'unknown',
+            idx = 0,
+            seed = 0,
+            binary = None,
+            uvm_log = None,
+            rtl_trace = None,
+            rtl_trace_csv = None,
+            iss_trace = None,
+            iss_trace_csv = None,
+            comparison_log = None,
+            passed = False,
+            failure_message = err
+    )
 
+def box_comment(line: str) -> str:
+    hr = '#' * 80
+    return hr + '\n# ' + line + '\n' + hr
 
-def parse_log(path: str) -> bool:
-    first_line = '(empty file)'
-    with open(path, 'r', encoding='UTF-8') as log:
-        for line in log:
-            first_line = line.rstrip()
-            break
+def gen_summary_line(passing_tests: List[TestRunResult], failing_tests:
+        List[TestRunResult]) -> str:
+    '''Generate a string summarising test results'''
+    total_tests = len(passing_tests) + len(failing_tests)
+    pass_pct = (len(passing_tests) / total_tests) * 100
 
-    if first_line == 'PASS':
-        return True
-    if first_line.startswith('FAIL'):
-        return False
+    return f'{pass_pct:0.2f}% PASS {len(passing_tests)} PASSED, ' \
+           f'{len(failing_tests)} FAILED'
 
-    raise RuntimeError('Strange first line ({!r})'.format(first_line))
+def gen_test_run_result_text(test_run_result: TestRunResult) -> str:
+    '''Generate a string describing a TestRunResult including the failure
+    message if the test did not pass'''
+    test_name_idx = f'{test_run_result.name}.{test_run_result.seed}'
+    test_underline = '-' * len(test_name_idx)
 
+    info_text = f'{test_name_idx}\n{test_underline}\n'
 
-def dump_log(path: str, dest: TextIO) -> None:
-    print('\nLog at {}:'.format(path), file=dest)
-    with open(path, 'r') as fd:
-        for line in fd:
-            dest.write('> ' + line)
+    if (test_run_result.binary):
+        info_text += f'Test binary: {test_run_result.binary}\n'
 
+    if (test_run_result.uvm_log):
+        info_text += f'UVM log: {test_run_result.uvm_log}\n'
 
-def box_comment(line: str) -> str:
-    hr = '#' * 80
-    return hr + '\n# ' + line + '\n' + hr
+    if (test_run_result.rtl_trace):
+        info_text += f'RTL trace: {test_run_result.rtl_trace}\n'
+
+    if (test_run_result.iss_trace):
+        info_text += f'ISS trace: {test_run_result.iss_trace}\n'
+
+    if (test_run_result.comparison_log):
+        info_text += f'Comparison log: {test_run_result.comparison_log}\n'
+
+    if (test_run_result.passed):
+        info_text += f'[PASSED]\n'
+    else:
+        info_text += f'{test_run_result.failure_message}\n'
+
+    return info_text
+
+def output_results_text(passing_tests: List[TestRunResult], failing_tests:
+        List[TestRunResult], dest: TextIO):
+    '''Write results in text form to dest'''
+
+    if failing_tests:
+        dest.write(box_comment('Details of failing tests'))
+        dest.write('\n')
```

Comment:
It might be simpler to use `print(something, dest=dest)` instead of the double `write` calls here. (Something that Philipp taught me!)